File: react-native-pytorch-core/android/src/main/java/org/pytorch/rn/core/ml/processing/unpacker/BertDecodeQAAnswerUnpacker.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package org.pytorch.rn.core.ml.processing.unpacker;

import com.facebook.react.bridge.WritableMap;
import java.io.IOException;
import java.util.Arrays;
import java.util.Map;
import org.json.JSONException;
import org.json.JSONObject;
import org.pytorch.IValue;
import org.pytorch.rn.core.ml.processing.BertTokenizer;
import org.pytorch.rn.core.ml.processing.PackerContext;
import org.pytorch.rn.core.ml.processing.Unpacker;

public class BertDecodeQAAnswerUnpacker implements Unpacker {

  private static final String CONTEXT_BERT_TOKENIZER_KEY = "bert_tokenizer";
  private static final String VOCABULARY_BERT_KEY = "vocabulary_bert";

  @Override
  public void unpack(
      IValue ivalue, JSONObject jobject, WritableMap result, PackerContext packerContext)
      throws JSONException, IOException {
    final Map<String, IValue> map = ivalue.toDictStringKey();
    float[] startLogits = map.get("start_logits").toTensor().getDataAsFloatArray();
    float[] endLogits = map.get("end_logits").toTensor().getDataAsFloatArray();

    final int startIdx = argmax(startLogits);
    final int endIdx = argmax(endLogits);
    // Return null (i.e., no answer found) if start index is outside the lower bounds of the tokens
    // or if start index is the same as the end index.
    if (startIdx < 0 || startIdx == endIdx) {
      return;
    }
    long[] tokenIds = (long[]) packerContext.get("token_ids");
    if (tokenIds == null) {
      throw new IllegalStateException("Expected 'token_ids' in packerContext");
    }

    final String answer =
        getBertTokenizer(packerContext).decode(Arrays.copyOfRange(tokenIds, startIdx, endIdx + 1));

    final String key = jobject.getString("key");
    result.putString(key, answer);
  }

  private int argmax(float[] array) {
    float max = -Float.MAX_VALUE;
    int ret = -1;
    for (int i = 0; i < array.length; ++i) {
      if (array[i] > max) {
        ret = i;
        max = array[i];
      }
    }
    return ret;
  }

  private BertTokenizer getBertTokenizer(PackerContext packerContext)
      throws JSONException, IOException {
    BertTokenizer bertTokenizer = (BertTokenizer) packerContext.get(CONTEXT_BERT_TOKENIZER_KEY);
    if (bertTokenizer == null) {
      bertTokenizer = new BertTokenizer(packerContext.specSrcJson.getString(VOCABULARY_BERT_KEY));
      packerContext.store(CONTEXT_BERT_TOKENIZER_KEY, bertTokenizer);
    }

    return bertTokenizer;
  }
}

File: react-native-pytorch-core/ios/ML/Processing/Unpacker/BertDecodeQAAnswerUnpacker.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation
import SwiftyJSON

class BertDecodeQAAnswerUnpacker: Unpacker {
    func unpack(ivalue: IValue, modelSpec: JSON, result: inout [String: Any], packerContext: PackerContext) throws {
        let unpack = modelSpec["unpack"]

        guard let key = unpack["key"].string else {
            throw BaseIValuePackerError.missingKeyParam
        }

        guard let bertTokenizer = packerContext.get(key: "bert_tokenizer") as? BertTokenizer,
              let tokenIds = packerContext.get(key: "token_ids") as? [Int] else {
            throw BaseIValuePackerError.decodeBertError
        }

        guard let dict = ivalue.toDictStringKey() as [String: IValue]? else {
            throw BaseIValuePackerError.decodeBertError
        }

        guard let startLogitTensor = dict["start_logits"]?.toTensor(),
              let endLogitTensor = dict["end_logits"]?.toTensor() else {
            throw BaseIValuePackerError.decodeBertError
        }

        let startLogits = startLogitTensor.getDataAsArrayBert()
        let endLogits = endLogitTensor.getDataAsArrayBert()
        let startIdx = argmax(array: startLogits, dtype: startLogitTensor.dtype)
        let endIdx = argmax(array: endLogits, dtype: endLogitTensor.dtype)

        // Return null (i.e., no answer found) if start index is outside the
        // lower bounds of the tokens or if start index is the same as the end
        // index.
        if startIdx < 0 || startIdx == endIdx {
            // Setting the answer nil is not necessarily required, but it is
            // explicit that the returned answer is nil (undefined in JS)
            result[key] = nil
            return
        }

        if startIdx > endIdx {
            throw BaseIValuePackerError.decodeBertError
        }

        let tokenIdRange = Array(tokenIds[startIdx...endIdx])
        do {
            let text = try bertTokenizer.decode(tokenIds: tokenIdRange)
            result[key] = text
        } catch {
            throw BaseIValuePackerError.decodeBertError
        }
    }

    private func argmax(array: [Any]?, dtype: PTMTensorType?) -> Int {
        guard let array = array, let dtype = dtype else {
            return 0
        }

        switch dtype {
        case .float:
            var max = -MAXFLOAT
            var ret = -1
            for (idx, item) in array.enumerated() {
                guard let num = item as? Float32 else {
                    continue
                }
                if num > max {
                  ret = idx
                  max = num
                }
            }
            return ret
        default:
            return 0
        }
    }
}
